File: Algorithm/Run Length Encoding/leetcode/38. Count and Say/Claude/Count-and-Say.py
# Python版のcount-and-say最適化ソリューションを提供します：## 🐍 **Python実装の特徴と最適化ポイント**
# Runtime
# 7
# ms
# Beats
# 75.02%
# Memory
# 18.24
# MB
# Beats
# 9.11%
# ### **1. 基本実装 (`Solution`)**
# - **型アノテーション**: 全てのパラメータと戻り値に型を明示
# - **Pylance対応**: 型エラーを回避する完全な型指定
# - **メモリ最適化**: `List[str]`を使用して文字列連結を効率化

# ### **2. 最適化実装 (`SolutionOptimized`)**
# - **高速RLE**: while文を使った効率的な文字カウント
# - **インデックス最適化**: 不要な文字比較を削減
# - **LeetCode最適**: 処理時間とメモリ使用量を最小化

# ### **3. メモ化実装 (`SolutionMemoized`)**
# - **動的プログラミング**: 計算済み結果をキャッシュ
# - **再帰最適化**: 大きなn値に対して効率的
# - **メモリトレードオフ**: 時間効率とメモリ使用量のバランス

# ## 📊 **パフォーマンス分析**

# | 実装版 | 時間計算量 | 空間計算量 | 特徴 |
# |--------|------------|------------|------|
# | Solution | O(m×k) | O(m) | 標準的、安定 |
# | SolutionOptimized | O(m×k) | O(m) | 高速、LeetCode最適 |
# | SolutionMemoized | O(m×k) | O(k×m) | 大きなn値に効率的 |

# ## 🔧 **Pylance型エラー対策**

# 1. **完全な型アノテーション**: 全変数に型を明示
# 2. **適切なimport**: `from typing import List` 
# 3. **メソッド型指定**: 戻り値型を明確に指定
# 4. **辞書型指定**: `dict[int, str]` で型安全性確保

# ## 🚀 **LeetCode最適化ポイント**

# 1. **文字列連結**: `''.join(result)` でO(n)効率
# 2. **リスト使用**: 中間結果をリストで保持
# 3. **早期終了**: 不要な処理をスキップ
# 4. **メモリ効率**: 必要最小限のデータ構造使用

# この実装は、Python 3.11.4環境とLeetCodeの要求を満たしつつ、型安全性とパフォーマンスの両方を確保しています。

from typing import List

class Solution:
    def countAndSay(self, n: int) -> str:
        """
        Count and Say数列のn番目の要素を返すメソッド
        
        Args:
            n (int): 取得したい数列の位置（1以上30以下）
            
        Returns:
            str: n番目のcount-and-say数列の文字列
            
        Time Complexity: O(m * k) - m: 各ステップの文字列長, k: ステップ数
        Space Complexity: O(m) - 各ステップで生成される文字列の長さ
        """
        # ベースケース: n=1の場合は"1"を返す
        current: str = "1"
        
        # n=1の場合はそのまま返す
        if n == 1:
            return current
        
        # 2からnまで反復的に数列を構築
        # for i in range(2, n + 1):
        for _ in range(2, n + 1):
            current = self._run_length_encode(current)
        
        return current
    
    def _run_length_encode(self, s: str) -> str:
        """
        文字列のRun-Length Encodingを行うヘルパーメソッド
        
        Args:
            s (str): エンコードする文字列
            
        Returns:
            str: エンコード後の文字列
            
        Time Complexity: O(len(s)) - 文字列の長さに比例
        Space Complexity: O(len(s)) - 結果文字列のサイズ
        """
        if not s:
            return ""
        
        result: List[str] = []  # メモリ効率のためリストを使用
        count: int = 1
        current_char: str = s[0]
        
        # 文字列を左から右にスキャン
        for i in range(1, len(s)):
            if s[i] == current_char:
                # 同じ文字が続く場合はカウントを増加
                count += 1
            else:
                # 異なる文字に変わった場合、カウントと文字を結果に追加
                result.append(str(count))
                result.append(current_char)
                current_char = s[i]
                count = 1
        
        # 最後の文字グループを追加
        result.append(str(count))
        result.append(current_char)
        
        # リストを結合して文字列として返す（メモリ効率的）
        return ''.join(result)


# 追加の最適化版（より高速な実装）
class SolutionOptimized:
    def countAndSay(self, n: int) -> str:
        """
        最適化されたCount and Say実装
        
        Args:
            n (int): 取得したい数列の位置（1以上30以下）
            
        Returns:
            str: n番目のcount-and-say数列の文字列
        """
        current: str = "1"
        
        for _ in range(n - 1):
            current = self._fast_rle(current)
        
        return current
    
    def _fast_rle(self, s: str) -> str:
        """
        高速なRun-Length Encoding実装
        
        Args:
            s (str): エンコードする文字列
            
        Returns:
            str: エンコード後の文字列
        """
        if not s:
            return ""
        
        result: List[str] = []
        i: int = 0
        length: int = len(s)
        
        while i < length:
            current_char: str = s[i]
            count: int = 1
            
            # 同じ文字を連続してカウント
            while i + count < length and s[i + count] == current_char:
                count += 1
            
            # カウントと文字を追加
            result.append(str(count))
            result.append(current_char)
            
            # 次のグループに移動
            i += count
        
        return ''.join(result)


# メモ化を利用した実装（大きなnに対して効率的）
class SolutionMemoized:
    def __init__(self) -> None:
        """コンストラクタ - メモ化用の辞書を初期化"""
        self.memo: dict[int, str] = {1: "1"}
    
    def countAndSay(self, n: int) -> str:
        """
        メモ化を使用したCount and Say実装
        
        Args:
            n (int): 取得したい数列の位置（1以上30以下）
            
        Returns:
            str: n番目のcount-and-say数列の文字列
        """
        if n in self.memo:
            return self.memo[n]
        
        # 再帰的にn-1の結果を取得してRLEを適用
        prev_result: str = self.countAndSay(n - 1)
        current_result: str = self._run_length_encode(prev_result)
        
        # 結果をメモ化
        self.memo[n] = current_result
        return current_result
    
    def _run_length_encode(self, s: str) -> str:
        """
        Run-Length Encodingヘルパーメソッド
        
        Args:
            s (str): エンコードする文字列
            
        Returns:
            str: エンコード後の文字列
        """
        if not s:
            return ""
        
        result: List[str] = []
        count: int = 1
        current_char: str = s[0]
        
        for char in s[1:]:
            if char == current_char:
                count += 1
            else:
                result.extend([str(count), current_char])
                current_char = char
                count = 1
        
        result.extend([str(count), current_char])
        return ''.join(result)

File: Algorithm/Run Length Encoding/leetcode/38. Count and Say/Claude/Count-and-Say.ts
// TypeScriptでの最適化されたcount-and-say解法を提供します：TypeScript版の最適化されたcount-and-say解法の特徴：
// Runtime
// 1
// ms
// Beats
// 95.97%
// Memory
// 57.86
// MB
// Beats
// 66.30%

// 1. **厳密な型定義**: すべての変数とパラメータに適切な型アノテーションを追加し、TypeScript 5.1の型チェック機能を活用しています。

// 2. **関数型アプローチ**: `var`ではなく`function`宣言を使用し、よりクリーンで読みやすいコードになっています。

// 3. **パフォーマンス最適化**:
//    - メモリ効率: 文字列の連結は効率的に行われ、不要なオブジェクト生成を避けています
//    - 時間効率: O(総文字数)の線形時間で処理されます

// 4. **型安全性**: TypeScriptの静的型チェックにより、実行時エラーのリスクを最小限に抑えています。

// 5. **LeetCode最適化**:
//    - 余分な配列やオブジェクトを使わない
//    - インライン文字列操作で高速化
//    - 必要最小限のメモリ使用量

// この実装は、LeetCodeの制約とTypeScript環境において最高のパフォーマンスを発揮するように設計されています。

/**
 * Count and Say数列のn番目の要素を返す関数
 * @param n - 取得したい数列の位置（1以上30以下）
 * @returns n番目のcount-and-say数列の文字列
 */
function countAndSay(n: number): string {
    // ベースケース: n=1の場合は"1"を返す
    let current: string = '1';

    // n=1の場合はそのまま返す
    if (n === 1) return current;

    // 2からnまで反復的に数列を構築
    for (let i: number = 2; i <= n; i++) {
        current = runLengthEncode(current);
    }

    return current;
}

/**
 * 文字列のRun-Length Encodingを行う関数
 * @param str - エンコードする文字列
 * @returns エンコード後の文字列
 */
function runLengthEncode(str: string): string {
    let result: string = '';
    let count: number = 1;
    let currentChar: string = str[0];

    // 文字列を左から右にスキャン
    for (let i: number = 1; i < str.length; i++) {
        if (str[i] === currentChar) {
            // 同じ文字が続く場合はカウントを増加
            count++;
        } else {
            // 異なる文字に変わった場合、カウントと文字を結果に追加
            result += count + currentChar;
            currentChar = str[i];
            count = 1;
        }
    }

    // 最後の文字グループを追加
    result += count + currentChar;

    return result;
}
